api/base/authentication/backends.py

```diff
@@ -5,8 +5,8 @@
 # https://docs.djangoproject.com/en/1.8/topics/auth/customizing/
 class ODMBackend(ModelBackend):
 
-    def authenticate(self, username=None, password=None):
-        return get_user(email=username, password=password) or None
+    def authenticate(self, request, username=None, password=None, email=None):
```

Comment:
example in 3.2 guide https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#writing-an-authentication-backend still has:

```python
class MyBackend(BaseBackend):
    def authenticate(self, request, username=None, password=None):
        ...
```

In addition, if we try to match the signature, should it be the following instead since this `email=None` can be anything like `foo=bar`.

```python
def authenticate(self, request, username=None, password=None, **kwargs):
    return get_user(email=username, password=password) or None
```

My admin still have problems locally but it is probably a local issue.

Reply:
Oh, we inherits from `ModelBackend` (instead of `BaseBackend` in the doc example), of which the signature is different.

```python
def authenticate(self, request, username=None, password=None, **kwargs):
    if username is None:
        username = kwargs.get(UserModel.USERNAME_FIELD)
    if username is None or password is None:
        return
    try:
        user = UserModel._default_manager.get_by_natural_key(username)
    except UserModel.DoesNotExist:
        # Run the default password hasher once to reduce the timing
        # difference between an existing and a nonexistent user (#20760).
        UserModel().set_password(password)
    else:
        if user.check_password(password) and self.user_can_authenticate(user):
            return user
```

Reply:
Forgot to mention the references

* [1.11 `ModelBackend`](https://github.com/django/django/blob/stable/1.11.x/django/contrib/auth/backends.py#L9-L25)
* [3.2 `ModelBackend`](https://github.com/django/django/blob/stable/3.2.x/django/contrib/auth/backends.py#L31-L49)